Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp

```diff
@@ -28,26 +28,154 @@
 
 #include "GPUAdapter.h"
 #include "GPUCanvasConfiguration.h"
+#include "RenderBox.h"
+#include <wtf/IsoMallocInlines.h>
+
+#if PLATFORM(COCOA)
+#include "WebGPUSurfaceCocoa.h"
+#endif
 
 namespace WebCore {
 
+static IntSize getCanvasSizeAsIntSize(const GPUCanvasContext::CanvasType& canvas)
+{
+    return WTF::switchOn(canvas, [](const RefPtr<HTMLCanvasElement>& htmlCanvas) -> IntSize {
+        return { static_cast<int>(htmlCanvas->width()), static_cast<int>(htmlCanvas->height()) };
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> IntSize {
+        return { static_cast<int>(offscreenCanvas->width()), static_cast<int>(offscreenCanvas->height()) };
+    }
+#endif
+    );
+}
+
+static bool platformSupportsWebGPUSurface()
+{
+#if PLATFORM(COCOA)
+    return true;
+#else
+    return false;
+#endif
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(GPUCanvasContext);
+
+std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
+{
+    if (!platformSupportsWebGPUSurface())
+        return nullptr;
+
+    auto context = std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
+    context->suspendIfNeeded();
+    return context;
+}
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+{
+    ASSERT(platformSupportsWebGPUSurface());
+}
+
+void GPUCanvasContext::reshape(int width, int height)
+{
+    if (!m_currentConfiguration)
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create({ width, height }, m_currentConfiguration->convertToBacking());
+#else
+    // This code should not be reached, because GPUCanvasContext::create prevents creating a new
+    // GPUCanvasContext when platformSupportsWebGPUSurface() is false.
+    UNUSED_PARAM(width);
+    UNUSED_PARAM(height);
+    ASSERT_NOT_REACHED();
+#endif
+}
+
 GPUCanvasContext::CanvasType GPUCanvasContext::canvas()
 {
-    return { };
+    return htmlCanvas();
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration configuration)
```

Comment:
Why copy?

Reply:
Looks like the intent is to copy into `m_currentConfiguration`. This should probably be an r-value reference instead for clarity, I will change it, however it will probably end up having the same effect:

so:
```
void GPUCanvasContext::configure(GPUCanvasConfiguration&&)
```